exec/java-exec/src/main/java/org/apache/drill/exec/store/pcap/decoder/TcpSession.java

```diff
@@ -0,0 +1,334 @@
+        // data,
+        handshake.setFin();
+      } else if (handshake.getCurrentSessionState() == TcpHandshake.State.CLOSE_WAIT) {
+
+      }
+      if (p.getAckFlag() && p.getFinFlag()) {
+        handshake.setAck();
+        handshake.finAck = true;
+      }
+    }
+
+    // Augment the packet counter
+    packetCount++;
+
+    if (p.isCorrupt()) {
+      hasCorruptedData = true;
+    }
+
+    // Add the start and ending time stamp.  The packets are not necessarily received in order, so we have to check the timestamps this way
+    if (p.getTimestampMicro() < startTime) {
+      startTime = p.getTimestamp();
+    }
+
+    if (p.getTimestampMicro() > endTime) {
+      endTime = p.getTimestamp();
+    }
+
+    // Close the session if the closing handshake is complete
+    if (handshake.isClosed()) {
+      closeSession();
+    }
+  }
+
+  /**
+   * This function returns true if the TCP session has been established, false if not.
+   * @return True if the session has been established, false if not.
+   */
+  public boolean connectionEstablished() {
+    return handshake.isConnected();
+  }
+
+  public boolean connectionClosed() {
+    return handshake.isClosed();
+  }
+
+  public void closeSession() {
+    logger.debug("Closing session {}", sessionID);
+    // Write the sent and received data
+    sessionLength = endTime - startTime;
+
+    /* The sent and received bytes cannot be written until the session is closed.
+    * Upon receipt of the FIN->FIN/ACK handshake, write everything.
+    *
+    * Since it cannot be assumed that the packets were received in the correct order, we must:
+    * 1. Sort them by TCP Sequence Number
+    * 2. Write the data to the respective byte array
+    */
+
+    Collections.sort(packetsFromSender);
+    Collections.sort(packetsFromReceiver);
+
+    sentData = new byte[sentDataSize];
+    receivedData = new byte[receivedDataSize];
+
+    byte[] dataFromPacket;
+    int dataOffset = 0;
+    // Now that the lists are sorted, add packet data to the lists
+    for (int i = 0; i < packetsFromSender.size(); i++) {
+      // Get the packet;
+      Packet p = packetsFromSender.get(i);
+      dataFromPacket = p.getData();
+      if (dataFromPacket != null) {
+        for (int j = 0; j < dataFromPacket.length; j++) {
+          sentData[dataOffset] = dataFromPacket[j];
+          dataOffset++;
+        }
+      }
+    }
+
+    dataOffset = 0;
+    for (int i = 0; i < packetsFromReceiver.size(); i++) {
+      // Get the packet;
+      Packet p = packetsFromReceiver.get(i);
+      dataFromPacket = p.getData();
+      if (dataFromPacket != null) {
+        for (int j = 0; j < dataFromPacket.length; j++) {
+          receivedData[dataOffset] = dataFromPacket[j];
+          dataOffset++;
+        }
+      }
+    }
+  }
+
+  @Override
+  public boolean equals (Object o) {
```

Comment:
If you override equals, you should override and hashcode, plus write more proper squash with instance of check first.

Reply:
I'm not sure why I originally overwrote the method, but I removed it.